x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/blob/BlobStoreCacheMaintenanceService.java

```diff
@@ -237,4 +392,293 @@ public void onFailure(Exception e) {
+                        }
+
+                        @Override
+                        public void onFailure(Exception e) {
+                            if (TransportActions.isShardNotAvailableException(e)) {
+                                complete(maintenanceTask, null);
+                            } else {
+                                complete(maintenanceTask, e);
+                            }
+                        }
+                    });
+                    return;
+                }
+
+                final String pitId = pointIntTimeId;
+                assert Strings.hasLength(pitId);
+
+                if (searchResponse == null) {
+                    final SearchSourceBuilder searchSource = new SearchSourceBuilder();
+                    searchSource.trackScores(false);
+                    searchSource.sort("_shard_doc");
+                    searchSource.size(batchSize);
+                    final PointInTimeBuilder pointInTime = new PointInTimeBuilder(pitId);
+                    searchSource.pointInTimeBuilder(pointInTime);
+                    pointInTime.setKeepAlive(keepAlive);
+                    final SearchRequest searchRequest = new SearchRequest();
+                    searchRequest.source(searchSource);
+                    if (searchAfter != null) {
+                        searchSource.searchAfter(searchAfter);
+                    }
+                    clientWithOrigin.execute(SearchAction.INSTANCE, searchRequest, new ActionListener<>() {
+                        @Override
+                        public void onResponse(SearchResponse response) {
+                            maintenanceTask.total.compareAndSet(0L, response.getHits().getTotalHits().value);
+                            maintenanceTask.searchResponse = response;
+                            maintenanceTask.searchAfter = null;
+                            executeNext(maintenanceTask);
+                        }
+
+                        @Override
+                        public void onFailure(Exception e) {
+                            complete(maintenanceTask, e);
+                        }
+                    });
+                    return;
+                }
+
+                final SearchHit[] searchHits = searchResponse.getHits().getHits();
+                if (searchHits != null && searchHits.length > 0) {
+                    final ClusterState state = clusterService.state();
+                    final BulkRequest bulkRequest = new BulkRequest();
+
+                    final TimeValue retention = periodicTaskRetention;
+                    final Instant expirationTime = Instant.ofEpochMilli(threadPool.absoluteTimeInMillis())
+                        .minus(retention.duration(), retention.timeUnit().toChronoUnit());
+
+                    // compute the list of existing searchable snapshots once
+                    Map<String, Set<String>> knownSnapshots = existingSnapshots;
```

Comment:
I think we can make this a field to avoid collecting this for every search round?

Reply:
I think it was already computed once. But I reworked this a bit in 15ca8cdd9394fc5e1c9d2c9abec84525a64c4b13 where the expiration time, snapshots and repositories are all computed once.